File: 146-lru-cache/lru-cache.cpp
struct Node{
    int key, val;
    Node* prev;
    Node* next;
    Node(int k, int v): key(k), val(v), prev(nullptr), next(nullptr){};
};

class LRUCache {
public:
    int cap;
    unordered_map<int, Node*> mp;
    Node* head = new Node(-1,-1);
    Node* tail = new Node(-1,-1);
    LRUCache(int capacity) {
        cap = capacity;
        head->next = tail;
        tail->prev = head;
    }
    



    void addNodeAtTail(Node* node){
        Node* temp;
        temp = tail->prev;
        temp->next = node;
        node->next = tail;
        node->prev = temp;
        tail->prev = node;
    }

    void removeNode(Node* node){
        Node* temp1 = node->prev;
        Node* temp2 = node->next;
        temp1->next = temp2;
        temp2->prev = temp1;
    }

    int get(int key) {
        if(mp.find(key)!= mp.end()){
            removeNode(mp[key]);
            addNodeAtTail(mp[key]);
            return mp[key]->val;
        } else{
            return -1;
        }  
    }
    
    void put(int key, int value) {
        if(mp.find(key)!= mp.end()){
            mp[key]->val = value;
            Node* updatedNode = mp[key];
            removeNode(updatedNode);
            addNodeAtTail(updatedNode);
        }else {
            Node* nodeToAdd = new Node(key, value);
            if(cap!=0){
                addNodeAtTail(nodeToAdd);
                mp.insert({key, nodeToAdd});
                cap--;
            }else{
                mp.erase(head->next->key);
                removeNode(head->next);
                addNodeAtTail(nodeToAdd);
                mp.insert({key, nodeToAdd});
            } 
        }     
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146-lru-cache/lru-cache.py
class ListNode:
        def __init__(self, key=0, val=0, l=None, r=None):
            self.key = key
            self.val = val
            self.l=l
            self.r=r

class LRUCache:

    def __init__(self, capacity: int):
        self.map = {}
        self.capacity = capacity
        self.head = ListNode(-1, -1)
        self.tail = ListNode(-1,-1)
        self.head.r = self.tail
        self.tail.l = self.head
        

    def get(self, key: int) -> int:
        if key not in self.map:
            return -1
        
        node = self.map[key]
        self.remove_node(node)
        self.insert_new_node_before_tail(node, self.tail)
        return node.val 

    def remove_node(self, root):
        prev = root.l
        next = root.r
        prev.r = next
        root.r = None
        next.l = prev
        root.l = None

    def insert_new_node_before_tail(self, node, tail):
        prev = tail.l
        prev.r = node
        node.l = prev
        node.r = tail
        tail.l = node    

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            self.map[key].val = value
            self.remove_node(self.map[key])
            self.insert_new_node_before_tail(self.map[key], self.tail)
        else:
            if len(self.map) >= self.capacity:
                del self.map[self.head.r.key]
                self.remove_node(self.head.r)
            new_node = ListNode(key, value)
            self.map[key] = new_node
            self.insert_new_node_before_tail(self.map[key], self.tail)
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)